### PR TITLE
Reference index

### PR DESCRIPTION
**Pull request by kaedroho**
This PR implments a new `ReferenceIndex` which records all references between models. This includes references from ForeignKeys, embeds/links in rich text and chooser blocks. The index is automatically updated on `post_save` and a `rebuild_references_index` management command is provided to rebuild it from scratch. The usage reports have been updated to use this index.

I have also removed the `WAGTAIL_USAGE_COUNT_ENABLED` setting, which was originally included to allow people to opt-in to usage count for performance reasons. Performance should be much better with the new approach so I think it should always be enabled.

### CODE REVIEW

File: wagtail/models/__init__.py [hunk 1]
```diff
@@ -4606,3 +4613,188 @@ class Meta:
...
+        )
+
+        # Delete removed references
+        deleted_references = set(existing_references.keys()) - references
+        cls.objects.filter(
+            id__in=[existing_references[reference] for reference in deleted_references]
+        ).delete()
+
+
+def get_object_usage(obj):
```
Comment (kaedroho) [inline wagtail/models/__init__.py:4780]: Delete this

File: wagtail/models/__init__.py [hunk 2]
```diff
@@ -4606,3 +4613,188 @@ class Meta:
...
+    # For example, say we have a StreamField called 'body' which has a struct block type
+    # called 'my_struct_block' that has a field called 'my_field'. If we extracted a
+    # reference from that field, the path would be set to the following:
+    #
+    # 'body.my_struct_block.my_field'
+    #
+    # We can then use this with the 'content_type' to find the original definition of
+    # the field block and display information to the user about where the reference was
+    # extracted from.
+    path = models.TextField()
```
Comment (kaedroho) [inline wagtail/models/__init__.py:4667]: Probably split this into two separate path fields:

 - `model_path` (For example `body.my_block.my_field`) - Used to find details of the field the reference came from, without having to retrieve the original object
 - `content_path (For example `body.bdc70d8b-e7a2-4c2a-bf43-2a3e3fcbbe86.my_field`) - Used to permalink the user to the specific block/field that contains the reference

File: wagtail/admin/templatetags/wagtailadmin_tags.py [hunk 3]
```diff
@@ -1079,3 +1080,14 @@ def human_readable_date(date, description=None):
...
+
+
+@register.simple_tag(takes_context=True)
+def get_edit_url(context, object):
+    user = None
+    request = getattr(context, "request", None)
+    if request:
+        user = request.user
+
+    url_finder = AdminURLFinder(user)
```
Comment (gasman) [inline wagtail/admin/templatetags/wagtailadmin_tags.py:1087]: I'm not really keen on us instantiating an `AdminURLFinder` on every item in the list - that's a lot of redundant reconstruction of lookup tables. More generally, I think creating a template tag for this is going to encourage more inefficient usage like this... Can we do the attaching of edit_url inside the view instead (using a single AdminURLFinder instance) and pass a tuple of (edit_url, reference) to the template?

File: wagtail/management/commands/rebuild_references_index.py [hunk 4]
```diff
@@ -29,16 +28,7 @@ def handle(self, **options):
...
-                    continue
-
-                # Don't check any models that have a parental key, references from these will be collected from the parent
-                if any(
-                    [
-                        isinstance(field, ParentalKey)
-                        for field in model._meta.get_fields()
-                    ]
-                ):
+                if not ReferenceIndex.model_is_indexible(model):
```
Comment (gasman) [inline wagtail/management/commands/rebuild_references_index.py:31]: I think "indexable" would be the more conventional spelling.

**Review by gasman — CHANGES_REQUESTED** [review]
Looking good! Still need to properly dig into the details of rich text and signal handlers, but here's some first-pass feedback in the meantime...

**Review by gasman — COMMENTED** [review]
I'm happy with this! The way that rich text rewriters get split up into `embed_rules` and `reference_extractors` feels a tad messy, but not sure how we'd do something better without overhauling the existing API, which would be out of scope for this - and keeping rewriters as the single "source of truth" about what embedded entities mean is definitely a good thing.

Just dropping in a small suggestion for making this more extensible.

**Review by gasman — APPROVED** [review]

**Review by gasman — CHANGES_REQUESTED** [review]
Rebased at https://github.com/gasman/wagtail/tree/kaedroho-reference-index with some fixes to the tests...

Just realised we need to consider permissions on the usage view, so that we're not leaking information about objects that the user isn't meant to be able to see. Unfortunately, doing this rigorously probably requires a lot more work, given that different models have different business rules about what a user should see - it would probably call for AdminURLFinder to be the source of truth for that.

Things get even more complicated if we want to filter the references queryset (which we'd need to do to list just the permitted items, while keeping pagination correct) - it would probably need something similar to our logging setup, with distinct tables for different permission models...

As a minimal acceptable fix, I suggest we treat edit permission as a best approximation to "things that should be shown in the usage report":

* For the 'usage' view on a given source object, require edit permission on that object
* When outputting the listing, any item with a null edit_url should output something like "(Private image)" instead of the object's `str` representation.

**Comment by gasman** [comment]
Thanks for adding the permission check @kaedroho!

Looks like we still need to address the other side - ensuring that the object listing doesn't show anything they're not supposed to see - but I'd be happy to merge this as is, and create a follow-up PR for that.

**Comment by gasman** [comment]
Afraid I've found some new issues on final testing:

* The usage count on snippets is unstyled and missing margins:
Screenshot 2022-09-30 at 14 15 27

It probably makes sense for this to appear in the Status side panel (which currently only appears for snippets with revisioning or draft status enabled, but could be made always-on)

On Bakerydemo, after running `rebuild_references_index`, the "Japan" snippet is still showing 0 uses despite being referenced from the Anpan breads page - investigating now.

**Comment by gasman** [comment]
> On Bakerydemo, after running rebuild_references_index, the "Japan" snippet is still showing 0 uses despite being referenced from the Anpan breads page - investigating now.

Got it - it's because `rebuild_references_index` is scanning the wagtailcore.Page model after breads.BreadPage. When retrieving the Anpan page as a plain Page, it naturally finds no references, and that empty list overwrites the real reference list that was stored under the same ID and content type.

**Comment by gasman** [comment]
> Got it - it's because `rebuild_references_index` is scanning the wagtailcore.Page model after breads.BreadPage. When retrieving the Anpan page as a plain Page, it naturally finds no references, and that empty list overwrites the real reference list that was stored under the same ID and content type.

This could be fixed for pages in particular by ensuring that we only ever index the `specific` instance. However, we'd strongly prefer to avoid special-casing the Page model, since in all other respects this implementation is agnostic of any Wagtail models. Since it acts on all Django models unless explicitly opted-out, we ought to ensure it works for all models using multi-table inheritance, and in the general case there may not be a way of obtaining a specific instance from a basic one (other than by scanning all installed models to find subclasses and doing a database lookup for each one).

The bug could also be handled within the `rebuild_references_index` command by keeping track of the already-indexed model / pk pairs, and skipping the indexing of any objects that we've previously indexed under a more specific subclass. However, this bug could also surface elsewhere any time we trigger a reindex on a model instance that isn't the most specific subclass (for example, by retrieving a `Page` instance, updating its title, and saving).

I haven't fully figured out a comprehensive fix yet, but I suspect it might require us to store the content type of the source object as two distinct fields, one for the base model and one for the specific model that we actually ran `ReferenceIndex.create_or_update_for_object` against. Then, on any subsequent call to `ReferenceIndex.create_or_update_for_object`, potentially with an object in its less specific form, it's only allowed to overwrite records where the specific_content_type field corresponds to that object or a more general subclass. That way, any references relating to a field on a more specific subclass will be left alone. (Alternatively, `create_or_update_for_object` could be made to take into account the `model_path` field of ReferenceIndex, and skip over any records that refer to an unrecognised field on the assumption that it belongs to a more specific subclass - however, this risks leaving behind stale entries if the field in question has actually been dropped in a migration.)

**Comment by gasman** [comment]
Another issue - it doesn't seem to be picking up document links in rich text. Tested by editing the FooterText snippet on bakerydemo to add a page link, document link and image, then running the following from the shell:

```
>>> from wagtail.models import ReferenceIndex
>>> from bakerydemo.base.models import FooterText
>>> ft = FooterText.objects.first()
>>> list(ReferenceIndex._extract_references_from_object(ft))
[(1, '61', 'body.', 'body.'), (2, '52', 'body.', 'body.')]
```
The two returned references correspond to the page and image - there is no reference returned for the document link.

**Comment by gasman** [comment]
Superseded by #9279

**Pull request closed without merging**